src/lib/sort/sort.ts

```diff
@@ -0,0 +1,147 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Directive, EventEmitter, Input, Output} from '@angular/core';
+import {SortDirection} from './sort-direction';
+import {coerceBooleanProperty} from '../core';
+import {getMdSortDuplicateMdSortableIdError, getMdSortHeaderMissingIdError} from './sort-errors';
+
+export interface MdSortable {
+  id: string;
+  start: SortDirection;
+  disableClear: boolean;
+  reverseOrder: boolean;
+}
+
+export interface Sort {
+  active: string;
+  direction: SortDirection;
+}
+
+/** Container for MdSortables to manage the sort state and provide default sort parameters. */
+@Directive({
+  selector: '[mdSort], [matSort]',
+})
+export class MdSort {
+  /** Collection of all registered sortables that this directive manages. */
+  sortables = new Map<string, MdSortable>();
+
+  /** The id of the most recently sorted MdSortable. */
+  @Input('mdSortActive') active: string;
+
+  /**
+   * The direction to set when an MdSortable is initially sorted.
+   * May be overriden by the MdSortable's sort start.
+   */
+  @Input('mdSortStart') start: SortDirection = 'ascending';
+
+  /** The sort direction of the currently active MdSortable. */
+  @Input('mdSortDirection') direction: SortDirection = '';
+
+  /**
+   * Whether to disable the user from clearing the sort by finishing the sort direction cycle.
+   * May be overriden by the MdSortable's disable clear input.
+   */
+  @Input('mdSortDisableClear')
+  get disableClear() { return this._disableClear; }
+  set disableClear(v) { this._disableClear = coerceBooleanProperty(v); }
+  private _disableClear: boolean;
+
+  /**
+   * Whether to reverse the default sort direction order cycle.
+   * May be overriden by the MdSortable's reverse order input.
+   */
+  @Input('mdSortReverseOrder')
+  get reverseOrder() { return this._reverseOrder; }
+  set reverseOrder(v) { this._reverseOrder = coerceBooleanProperty(v); }
+  private _reverseOrder: boolean;
+
+  /** Event emitted when the user changes either the active sort or sort direction. */
+  @Output() mdSortChange = new EventEmitter<Sort>();
+
+  /**
+   * Register function to be used by the contained MdSortables. Adds the MdSortable to the
+   * collection of MdSortables.
+   * @docs-private
```

Comment:
Hmm I wasn't thinking others needed it but people implementing their own `MdSortable` certainly would. Removed